crates/biome_js_analyze/src/lint/nursery/no_global_dirname_filename.rs

```diff
@@ -0,0 +1,257 @@
+use crate::{services::semantic::Semantic, JsRuleAction};
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, ActionCategory, FixKind, Rule, RuleDiagnostic,
+    RuleSource,
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_semantic::SemanticModel;
+use biome_js_syntax::{
+    global_identifier, AnyJsExpression, AnyJsName, AnyJsObjectMember, JsObjectExpression,
+    JsPropertyObjectMember, JsStaticMemberExpression, JsSyntaxKind, JsSyntaxToken,
+    JsVariableDeclarator,
+};
+use biome_rowan::{declare_node_union, AstSeparatedList, BatchMutationExt, TriviaPieceKind};
+
+declare_lint_rule! {
+    /// Disallow the use of `__dirname` and `__filename` in the global scope.
+    ///
+    /// They are [not available in ES modules](https://nodejs.org/api/esm.html#esm_no_filename_or_dirname).
+    /// Starting with Node.js 20.11, `import.meta.dirname` and `import.meta.filename` have been introduced in ES modules, providing identical functionality to `__dirname` and `__filename` in CommonJS (CJS).
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const dirname = __dirname;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const filename = __filename;
+    /// ```
+    ///
+    /// ``` js,expect_diagnostic
+    /// const foo = { __filename }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// if (__dirname.startsWith("/project/src/")) {}
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const dirname = import.meta.dirname
+    /// const filename = import.meta.filename
+    /// const foo = {__filename: import.meta.filename };
+    /// if (import.meta.dirname.startsWith("/project/src/")) {}
+    /// ```
+    ///
+    pub NoGlobalDirnameFilename {
+        version: "next",
+        name: "noGlobalDirnameFilename",
+        language: "js",
+        recommended: false,
+        sources: &[RuleSource::EslintUnicorn("prefer-module")],
+        fix_kind: FixKind::Safe,
+    }
+}
+
+declare_node_union! {
+    pub AnyGlobalDirnameFileName =
+        JsVariableDeclarator
+        | JsObjectExpression
+        | JsStaticMemberExpression
+}
+
+impl Rule for NoGlobalDirnameFilename {
+    type Query = Semantic<AnyGlobalDirnameFileName>;
+    type State = (JsSyntaxToken, String);
+    type Signals = Vec<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let model = ctx.model();
```

Comment:
This rule should check if we're evaluating a script file `.cjs`, and if so, the rule should bail because they are valid in a CommonJS file